services/placement/filter/mirrored.go

```
@@ -0,0 +1,435 @@
+
+package filter
+
+import (
+	"container/heap"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/services/placement"
+)
+
+var (
+	errNoValidMirrorInstance = errors.New("no valid instance for mirror placement in the candidate list")
+)
+
+type mirroredFilter struct {
+	opts services.PlacementOptions
+}
+
+func newMirroredFilter(opts services.PlacementOptions) placement.InstanceFilter {
+	return &mirroredFilter{opts: opts}
+}
+
+func (f *mirroredFilter) FilterInitialInstances(
+	candidates []services.PlacementInstance,
+	rf int,
+) ([]services.PlacementInstance, error) {
+	candidates, err := getValidCandidates(placement.NewPlacement(), candidates, f.opts)
+	if err != nil {
+		return nil, err
+	}
+
+	weightToHostMap, err := hostsGroupedByWeight(candidates)
+	if err != nil {
+		return nil, err
+	}
+
+	var pairs = make([][]services.PlacementInstance, 0, len(candidates))
+	for _, hosts := range weightToHostMap {
+		pairedHosts := pairHosts(hosts, rf)
+		if len(pairedHosts) == 0 {
+			continue
+		}
+
+		pairedInstances, err := pairInstances(pairedHosts)
+		if err != nil {
+			return nil, err
+		}
+
+		pairs = append(pairs, pairedInstances...)
+	}
+
+	if len(pairs) == 0 {
+		return nil, errNoValidMirrorInstance
+	}
+
+	return assignShardsToPairedInstances(pairs, placement.NewPlacement()), nil
+}
+
+func (f *mirroredFilter) FilterAddingInstances(
+	candidates []services.PlacementInstance,
+	p services.Placement,
+) ([]services.PlacementInstance, error) {
+	candidates, err := getValidCandidates(p, candidates, f.opts)
+	if err != nil {
+		return nil, err
+	}
+
+	weightToHostMap, err := hostsGroupedByWeight(candidates)
+	if err != nil {
+		return nil, err
+	}
+
+	var pairs = make([][]services.PlacementInstance, 0, len(candidates))
+	for _, hosts := range weightToHostMap {
+		pairedHosts := pairHosts(hosts, p.ReplicaFactor())
+		if len(pairedHosts) == 0 {
+			continue
+		}
+
+		pairs, err = pairInstances(pairedHosts[:1])
```

hmm, probably not, the only reason this could error out is when hosts have different ports available, which should never happen. Also when that happens we may want to fail early and debug what caused the issue(most likely the r2admin created bad instances).

Right now there is no logic that prefers weights, not sure if we want to prefer larger or smaller weight, but the logic could be added.